llvm/include/llvm/Analysis/TargetTransformInfo.h

```diff
@@ -1616,6 +1616,11 @@ class TargetTransformInfo {
       TTI::TargetCostKind CostKind = TTI::TCK_RecipThroughput,
       bool UseMaskForCond = false, bool UseMaskForGaps = false) const;
 
+  /// \return The cost of vp intrinsic vp.load.ff.
+  LLVM_ABI InstructionCost getFirstFaultLoadCost(
```

Comment:
What is the benefit of having a separate hook instead of just requesting the cost of the intrinsic?

Reply:
We pass info to getIntrinsicInstrCost via IntrinsicCostAttributes, but attributes like alignment aren’t supported there.

Reply:
getVPMemoryOpCost looks like the VP counterpart of getMemoryOpCost.
My read is that Opcode only encodes Instruction::Load vs Instruction::Store; it doesn’t select among VP intrinsics (Intrinsic::vp_load, ::vp_store, ::vp_load_ff).

Reply:
TIL that `getVPMemoryOpCost` actually exists. We don't use it in RISC-V, so it looks like it's dead. I've opened up a PR to remove it: https://github.com/llvm/llvm-project/pull/160838

Given that we're moving away from VP intrinsics in general, if SVE ever wants to support first-fault loads I presume we'll want to have a non-VP hook anyway?

Reply:
> TIL that getVPMemoryOpCost actually exists. We don't use it in RISC-V, so it looks like it's dead. I've opened up a PR to remove it: https://github.com/llvm/llvm-project/pull/160838

This is exactly why I would argue we should try to adress the underlying issue, i.e. make `getIntrinsicInstrCost` powerful enough to serve the use cases, so we don't have to add various very specialized hooks when they are for intrinsics.

@arcbbb would probably have been good to wait a bit with merging while there was still an ongoing discussion

Reply:
>  i.e. make getIntrinsicInstrCost powerful enough to serve the use cases, so we don't have to add various very specialized hooks when they are for intrinsics.

I think that's reasonable. We should probably also consolidate all the different memory costing hooks that can go into getIntrinsicInstrCost? E.g.

- getMaskedMemoryOpCost -> llvm.masked.load/store
- getGatherScatterOpCost ->  llvm.masked.gather/scatter
- getExpandCompressMemoryOpCost -> llvm.masked.expandload/llvm.masked.compressstore
- getStridedMemoryOpCost -> llvm.experimental.vp.strided.load/store

Reply:
@fhahn Thanks for flagging; I’ve reverted the change so we can discuss this without rushing.

Reply:
Thanks @arcbbb!

> I think that's reasonable. We should probably also consolidate all the different memory costing hooks that can go 
> into getIntrinsicInstrCost? E.g.
> 
>     getMaskedMemoryOpCost -> llvm.masked.load/store
>     getGatherScatterOpCost -> llvm.masked.gather/scatter
>     getExpandCompressMemoryOpCost -> llvm.masked.expandload/llvm.masked.compressstore
>     getStridedMemoryOpCost -> llvm.experimental.vp.strided.load/store

I think that would be an ideal outcome, but I am not sure how easy it would be to pass the required info through.

 Depending on that, it would probably be OK to add another hook like in this PR (at least `getVPMemoryOpCost` is gone now), but would be great to check if the general direction would be feasible.